c3g/freezeman
# Experiments (v3.3)

New PR as the result of the merging of the back-end WIP https://github.com/c3g/freezeman/pull/94 and the front-end WIP https://github.com/c3g/freezeman/pull/99  

TODO: 
- [x] Ingestion of the template through the front-end
- [x] Display Samples in a ExperimentRun page tab
- [x] Handling errors related to the Sample section of the Experiment template
  - [x] Enough volume left in the source samples (and it is not depleted) for the Experiment
  - [x] Source container barcode is found
  - [x] 2 samples in the same experiment container with the same position
- [x] Handling depletion of sample during ExperimentRun
- [ ] ExperimentRun resource test
- [ ] Cypress integration tests for the template, for the list page, and for different entity pages


TO VERIFY:
- [x] Verify for the volume used for a sample (should it be done in the ProcessMeasurement for the transfer VS the experiment) 
- [ ] The way the date and time properties are stored as JSON in PropertyValue

BONUS TASKS:
- [x] Add the Platform and the Instrument Type as columns and information for ExperimentRuns
- Improve the ExperimentRunResource logic to make it even more modular (i.e. row numbers for sections hardcoded for now)


ASSUMPTIONS:
- Samples related to an Experiment are fetched through their association to their Container. This means, that the Samples should not be transferred over to another Container after being used in an Experiment. Also, Container would not be re-used for another Experiment. If this logic is not true, we should change some of the calls to the API from the front-end.
